engine/src/test/java/com/datatorrent/stram/StreamingAppMasterServiceTest.java

```diff
@@ -0,0 +1,319 @@
+    AMRMClientImpl<ContainerRequest> client = mock(AMRMClientImpl.class);
+    AllocateResponse allocateResponse = mock(AllocateResponse.class);
+    when(client.allocate(0)).thenReturn(allocateResponse);
+
+    doReturn(response).when(client).registerApplicationMaster(any(String.class), any(int.class), any(String.class));
+
+    doAnswer(new Answer<List<ContainerStatus>>()
+    {
+      public List<ContainerStatus> answer(InvocationOnMock invocation)
+      {
+        return getListOfContainerStatuses();
+      }
+    }).when(allocateResponse).getCompletedContainersStatuses();
+
+    doAnswer(new Answer<List<Container>>()
+    {
+      public List<Container> answer(InvocationOnMock invocation)
+      {
+        return getListOfContainers();
+      }
+    }).when(allocateResponse).getAllocatedContainers();
+
+    appMaster.setAmRmClient(client);
+
+    mockedAppMaster = spy(appMaster);
+    LaunchContainerRunnable obj = mock(LaunchContainerRunnable.class);
+
+    doReturn(obj).when(mockedAppMaster).createLaunchContainerRunnable(any(Container.class), any(StreamingContainerAgent.class), any(ByteBuffer.class));
+
+    doReturn(resourceRequestor).when(mockedAppMaster).createResourceRequestor();
+    doReturn(true).when(mockedAppMaster).setupRMService(any(Configuration.class), any(int.class), any(ResourceRequestHandler.class));
+    when(mockedAppMaster.getConfig()).thenReturn(conf);
+  }
+
+  @After
+  public void teardown()
+  {
+  }
+
+  @Test
+  public void testBlacklistingOfFailedNodes() throws Exception
+  {
+    LogicalPlan dag = setupDag();
+    dag.setAttribute(LogicalPlan.MAX_CONSECUTIVE_CONTAINER_FAILURES, 3);
+
+    Thread thread = new Thread("LocalAMService")
+    {
+      @Override
+      public void run()
+      {
+        long startTms = System.currentTimeMillis();
+        long timeout = 50000L;
+        try {
+          while (mockedAppMaster.getBlacklistedNodes().isEmpty() && System.currentTimeMillis() - startTms < timeout) {
+            Thread.sleep(500);
+          }
+        } catch (InterruptedException ex) {
+          DTThrowable.rethrow(ex);
+        } finally {
+          mockedAppMaster.setAppDone(true);
+        }
+      }
+    };
+    thread.start();
+
+    mockedAppMaster.run();
+    thread.join();
+    Assert.assertEquals("Node1 should be blacklisted after 3 consecutive failures", "node1", mockedAppMaster.getBlacklistedNodes().get(0));
+    Assert.assertEquals("Node1 should have 3 failures marked", 3, mockedAppMaster.getFailedContainersMap().get("node1").get());
+  }
+
+  @Test
+  public void testBlacklistedNodesAreRemovedAfterTimeout() throws Exception
+  {
+    LogicalPlan dag = setupDag();
+    dag.setAttribute(LogicalPlan.BLACKLIST_REMOVAL_TIME, new Long(0));
+
+    final Thread thread = new Thread("LocalAMService")
+    {
+      @Override
+      public void run()
+      {
+        long startTms = System.currentTimeMillis();
+        long timeout = 50000L;
```

Comment:
They are actually not derived. Did trial runs and set these values. They can be changed if needed.